rules/downgrade-php73/src/Rector/List_/DowngradeListReferenceAssignmentRector.php

```diff
@@ -217,4 +273,40 @@ private function createAssignRefWithArrayDimFetch(
         $arrayDimFetch = new ArrayDimFetch($nestedExprVariable, $dim);
         return new AssignRef($assignVariable, $arrayDimFetch);
     }
+
+    /**
+     * @param (ArrayItem|null)[] $items
+     * @return ArrayItem[]
+     */
+    private function getItemsByRef(array $items, int $condition): array
+    {
+        /** @var ArrayItem[] */
+        return array_filter(array_map(
+            function (?ArrayItem $item) use ($condition): ?ArrayItem {
+                return $this->getItemByRefOrNull($item, $condition);
+            },
+            $items
+        ));
+    }
+
+    private function getItemByRefOrNull(?ArrayItem $arrayItem, int $condition): ?ArrayItem
+    {
+        if ($arrayItem === null) {
+            return null;
+        }
+        // Check if the item is a nested list/nested array destructuring
+        if ($arrayItem->value instanceof List_ || $arrayItem->value instanceof Array_) {
+            // Recursive call
+            /** @var List_|Array_ */
+            $nestedList = $arrayItem->value;
+            $hasItemByRef = false;
+            if ($condition === self::ALL) {
+                $hasItemByRef = $this->hasAllItemsByRef($nestedList->items);
+            } elseif ($condition === self::ANY) {
+                $hasItemByRef = $this->hasAnyItemByRef($nestedList->items);
+            }
+            return $hasItemByRef ? $arrayItem : null;
+        }
+        return $arrayItem->value instanceof Variable && $arrayItem->byRef ? $arrayItem : null;
+    }
```

Comment:
Btw, this starts to slow like a God class. I need to lower class complexity for PHPStan.

Could you decouple some logic you find most valuable to standalone services?